### PR TITLE
Bug: commander kill switch require 2 consecutive manual setpoints

### PR DESCRIPTION
**Pull request by dagar**

### CODE REVIEW

**Comment by mhkabir** [comment]
Why check for 2 specifically?

**Comment by dagar** [comment]
The important change here is that the bar for good RC is raised to 2 consecutive frames separated by less than the RC timeout and not just a single on time message. This prevents some potential weird cases if it's lost and then regained or a spurious message gets through. Since we're already keeping the current and previous manual control setpoints for other reasons it's easy to to require both for the lockdown (kill switch) activation. The additional delay isn't noticeable and it potentially stops a false positive.

**Comment by dagar** [comment]
This PR also eliminates "loss of RC" when calibrating.

File: src/modules/commander/commander.cpp [hunk 1]
```diff
@@ -2216,8 +2217,11 @@ Commander::run()
 		}
 
 		/* RC input check */
-		if (!status_flags.rc_input_blocked && sp_man.timestamp != 0 &&
-		    (hrt_elapsed_time(&sp_man.timestamp) < (rc_loss_timeout * 1_s))) {
+		if (!status_flags.rc_input_blocked
+			&& (sp_man.timestamp > _last_sp_man.timestamp)
+			&& ((sp_man.timestamp - _last_sp_man.timestamp) < (rc_loss_timeout * 1_s))
+			&& (hrt_elapsed_time(&sp_man.timestamp) < (rc_loss_timeout * 1_s))
```
Comment (LorenzMeier) [inline src/modules/commander/commander.cpp]: hrt_elapsed_time is expensive CPU load wise and should be only used where absolutely required.

File: src/modules/commander/commander.cpp [hunk 2]
```diff
@@ -2381,6 +2390,8 @@ Commander::run()
 			}
 		}
 
+		_last_sp_man = sp_man;
```
Comment (LorenzMeier) [inline src/modules/commander/commander.cpp]: Changing the position here could have unexpected side-effects.

**Comment by LorenzMeier** [comment]
@bkueng Could you please cross-test?

File: src/modules/commander/commander.cpp (hunk 1 above)
Comment (bkueng) [reply src/modules/commander/commander.cpp]: +1

File: src/modules/commander/commander.cpp (hunk 2 above)
Comment (bkueng) [reply src/modules/commander/commander.cpp]: Agreed, too risky for the release.

File: src/modules/commander/commander.cpp (hunk 1 above)
Comment (dagar) [reply src/modules/commander/commander.cpp]: I actually have timing numbers for a lot of these things if you're curious. An hrt_elapsed_time call takes a bit more than a microsecond, which for comparison is cheaper than most trig operations (single precision).

**Comment by dagar** [comment]
For reference I occasionally see things like this in logs from systems with no RC attached. It's likely a hardware issue that can be resolved, but I don't see the harm in also having slightly higher requirements for considering RC healthy.

image

File: src/modules/commander/commander.cpp (hunk 2 above)
Comment (dagar) [reply src/modules/commander/commander.cpp]: Alright, let's review this post v1.8.0 then. In some cases the way some of these features actually function has drifted from what I think may have been intended.

**Comment by dagar** [comment]
We need to revisit this, I've received more reports of accidental disarming from a single blip in RC.

https://github.com/PX4/Firmware/pull/9712

**Pull request closed without merging**